### ISSUE TITLE
Does not find network scanner which is found by non-flatpak'd Skanlite

### ISSUE THREAD

**Issue by Alexander-Wilms** [flathub/org.kde.skanlite#3]

**Comment by travier** [comment]
Can you try the nightly from https://community.kde.org/Guidelines_and_HOWTOs/Flatpak#Applications ? We've had success with those in the past but did not dig into the exact difference that made them work.

**Comment by Alexander-Wilms** [comment]
org.kde.skanlite//master didn't detect it either: https://bugs.kde.org/show_bug.cgi?id=458048

**Comment by kasars** [comment]
I think the reason is that you do not have the brother sane backend installed. You might need to install a package from brother support.

**Comment by travier** [comment]
Can you give us the list of packages related to sane from your distribution so that we can see what's missing in the Flatpak? Thanks!

**Comment by travier** [comment]
From Arch Linux, I can find: https://wiki.archlinux.org/title/SANE/Scanner-specific_problems#Brother but I'm not sure if or how we can redistribute those files with Flathub.

**Comment by Erick555** [comment]
if the scanner supports escl or wsd then it should be possible to use it with sane-airscan backend that was recently added here. Is this machine very old?

**Comment by Erick555** [comment]
Yes, but flatpak may not be aware of any of this so best chance is to try with driverless.

**Comment by hfiguiere** [comment]
maybe there should be extensions for SANE drivers? There are already a few Flatpak that uses SANE.

**Comment by hfiguiere** [comment]
> I had already installed the scanner driver using Brother's driver install tool and I can scan without problems using Skanpage, Skanlite or Simple Scan installed from openSUSE's RPM repository.

The Flatpak is snadboxed so it will not see anything you install with your distro. And never will.

**Comment by Erick555** [comment]
I don't see extension can scale when only brother have 4 different drivers, available at obscure urls, protected by EULA...

I think the airscan is becoming new standard even for native apps.

**Comment by hfiguiere** [comment]
Brother is not the only game in town.  EPSON also.

**Comment by Erick555** [comment]
Yeah its what I meant  - there are not only many vendors but multiple drivers per vendor. One size fits all extension would be a complete mess.

**Comment by mbugni** [comment]
I tried this app using my old Acer FlatbedScanner, which requires custom firmware to work.
I built skanlite locally from source:

```shell
$ flatpak-builder --user --verbose --install --install-deps-from=flathub --force-clean build org.kde.skanlite.json
```

So I configured my local `/etc/sane.d/snapscan.conf` to point the firmware under `/usr/share/sane/snapscan/firmware.bin`.
The scanner is properly detected inside the sandbox:

```shell
[📦 org.kde.skanlite ~]$ scanimage -L
device `v4l:/dev/video0' is a Noname HP TrueVision HD Camera: HP Tru virtual device
device `snapscan:libusb:001:012' is a Acer FlatbedScanner24 flatbed scanner
```

I launched skanlite and it works, both using Fedora 39 and Debian 12:

```shell
$ flatpak --user run org.kde.skanlite
```
![flatpak-skanlite](https://github.com/user-attachments/assets/885214c9-8134-4fee-a881-35eb652a550f)

I tested the scanner also with `xsane` from the host system, both with native and network driver.
From the host I can also see the scanner as a network device (because I configured saned):

```shell
native@host ~$ scanimage -L
device `v4l:/dev/video0' is a Noname HP TrueVision HD Camera: HP Tru virtual device
device `snapscan:libusb:001:012' is a Acer FlatbedScanner24 flatbed scanner
device `net:localhost:snapscan:libusb:001:012' is a Acer FlatbedScanner24 flatbed scanner
```

But from the sandbox the scanner doesn't work as network device, because the file `/etc/sane.d/net.conf` does not contain the `localhost` entry.
Basically it works for me, except for remote device driver.

**Comment by mbugni** [comment]
I managed to get the scanner working through the network.
First, override the SANE_CONFIG_DIR *inside* the sandbox:

```shell
[📦 org.kde.skanlite ~]$ export SANE_CONFIG_DIR="$XDG_CONFIG_HOME/sane.d:"
```
Then configure the `net` driver by overriding the `$HOME/.var/app/org.kde.skanlite/config/sane.d/net.conf` file.
So, from *ouside* the sandbox (in the host system) you can see:
```shell
native@host $ cat $HOME/.var/app/org.kde.skanlite/config/sane.d/net.conf 
# This is the net backend config file.

## saned hosts
# Each line names a host to attach to.
# If you list "localhost" then your backends can be accessed either
# directly or through the net backend.  Going through the net backend
# may be necessary to access devices that need special privileges.
localhost
```

From *inside* the sandbox (using the Flatpak runtime) you can see:
```shell
[📦 org.kde.skanlite ~]$ cat $XDG_CONFIG_HOME/sane.d/net.conf 
# This is the net backend config file.

## saned hosts
# Each line names a host to attach to.
# If you list "localhost" then your backends can be accessed either
# directly or through the net backend.  Going through the net backend
# may be necessary to access devices that need special privileges.
localhost
```

Now the scanner is visible as network device:
```shell
[📦 org.kde.skanlite ~]$ scanimage -L
device `v4l:/dev/video0' is a Noname HP TrueVision HD Camera: HP Tru virtual device
device `snapscan:libusb:001:031' is a Acer FlatbedScanner24 flatbed scanner
device `net:localhost:snapscan:libusb:001:031' is a Acer FlatbedScanner24 flatbed scanner
```

Now you can run Skanlite (*after* the SANE_CONFIG_DIR setup):
```shell
[📦 org.kde.skanlite ~]$ skanlite
```

@travier I can open a separate issue for this, because I think the SANE_CONFIG_DIR sholud be set by default, in order to allow users to customize drivers inside the sandbox.

**Comment by Erick555** [comment]
would it work when SANE_CONFIG_DIR points to non-existing path?

**Comment by mbugni** [comment]
> would it work when SANE_CONFIG_DIR points to non-existing path?

No issues at all:

```shell
[📦 org.kde.skanlite ~]$ export SANE_CONFIG_DIR="/very/fake/path:$XDG_CONFIG_HOME/sane.d:"

[📦 org.kde.skanlite ~]$ scanimage -L
device `v4l:/dev/video0' is a Noname HP TrueVision HD Camera: HP Tru virtual device
device `snapscan:libusb:001:014' is a Acer FlatbedScanner22 flatbed scanner
device `net:localhost:snapscan:libusb:001:014' is a Acer FlatbedScanner22 flatbed scanner
```

Note the final `:`. You can take a look at documentation:

```
SANE_CONFIG_DIR
This environment variable specifies the list of directories that
may contain the configuration file.  On *NIX systems, the direc-
tories are separated by a colon (`:'), under OS/2, they are sep-
arated by a semi-colon (`;').  If this variable is not set,  the
configuration  file  is  searched  in  two  default directories:
first, the current working directory (".") and then in  /usr/lo-
cal/etc/sane.d.   If  the value of the environment variable ends
with the directory separator character, then the default  direc-
tories  are searched after the explicitly specified directories.
For example, setting SANE_CONFIG_DIR to "/tmp/config:" would re-
sult in directories tmp/config, ., and /usr/local/etc/sane.d be-
ing searched (in this order).
```

**Comment by mbugni** [comment]
Another test:

```shell
[📦 org.kde.skanlite ~]$ export SANE_CONFIG_DIR="$XDG_CONFIG_HOME/sane.d:"
[📦 org.kde.skanlite ~]$ scanimage -L
device `v4l:/dev/video0' is a Noname HP TrueVision HD Camera: HP Tru virtual device
device `snapscan:libusb:001:015' is a Acer FlatbedScanner24 flatbed scanner
device `net:localhost:snapscan:libusb:001:015' is a Acer FlatbedScanner24 flatbed scanner

[📦 org.kde.skanlite ~]$ export SANE_CONFIG_DIR=":"
[📦 org.kde.skanlite ~]$ scanimage -L
device `v4l:/dev/video0' is a Noname HP TrueVision HD Camera: HP Tru virtual device
device `snapscan:libusb:001:015' is a Acer FlatbedScanner24 flatbed scanner
```

So the path matters.
